### PR TITLE
Add hbit module

### LINKED ISSUE #2606: LN code is based on unsupported trading pair
## Problem

When we wrote the LN code we used heth/halight as the swap protocols. It turns out we do not want to support this combination.

## Goal

Have a useful first LN implementation to base future work off of.

DoD:
- An passing e2e test that tests the happy path for a herc20/halight swap.

### PR DESCRIPTION
**Pull request by tcharding**
We have been trying to implement the newly designed split protocols for ages now; this PR is a change of approach. Instead of pushing forward with the LN stuff hoping that a nice split protocols implementation will fall out this PR adds the herc20/hbit protocol pair.
 
Add an `hbit` module under `swap_protocols/` with the following things in mind:
- Do not use rfc003 types, duplicate rfc003 code into `hbit` if needed.
- Try not to break any current working code.

This PR adds two new end points for both swaps of the trading pair hbit/herc20. Save of the created swap is fully implemented.  No retrieval for this data from the data base, also the REST actions are not done (these are heavy users of data retrieval so will be interesting to implement on top of the new database design, this can be done after this merges without hampering our effort).
  
Adds `unimplemeneted!` for all entrypoints to the networking layer. 

Please note, during duplication of the generic rfc003 code the Deploy state was copied, this has not been removed - prefer doing so separately after merge.

### CODE REVIEW

File: cnd/src/swap_protocols/hbit.rs [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    L: Clone + ledger::bitcoin::Bitcoin + ledger::bitcoin::Network,
+{
+    create_watcher::<_, _, _, _, htlc_location::Bitcoin, _, transaction::Bitcoin>(
+        connector.as_ref(),
+        ledger_states,
+        swap_id,
+        htlc_params,
+        Utc::now().naive_local(),
+    )
+    .instrument(tracing::error_span!("alpha_ledger", swap_id = %swap_id, role = %role))
```
Comment (thomaseizinger) [inline cnd/src/swap_protocols/hbit.rs:82]: Saw the same thing in Daniel's PR. I don''t know where this was copied from but at the time when I originally wrote these tracing statements, they made sense and this `alpha_ledger` was actually always true. This is no longer the case now.

As this needs some more thinking and probably refactoring to balance duplication and correctness, I would suggest to remove it for now.

File: cnd/src/swap_protocols/hbit.rs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+where
+    C: HtlcFunded<L, A, H, I, T>
+        + HtlcDeployed<L, A, H, I, T>
+        + HtlcRedeemed<L, A, H, I, T>
+        + HtlcRefunded<L, A, H, I, T>,
+    Deployed<H, T>: Clone,
+    Redeemed<T>: Clone,
+    Refunded<T>: Clone,
+{
+    let deployed = connector.htlc_deployed(&htlc_params, start_of_swap).await?;
```
Comment (thomaseizinger) [inline cnd/src/swap_protocols/hbit.rs:167]: Bitcoin HTLCs don't have a separate `deploy` step.

File: cnd/src/swap_protocols/hbit.rs [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+                return;
+            }
+        }
+    }
+}
+
+/// Returns a future that waits for events to happen on a ledger.
+///
+/// Each event is yielded through the controller handle (co) of the coroutine.
+async fn watch_ledger<C, L, A, H, I, T>(
```
Comment (thomaseizinger) [inline cnd/src/swap_protocols/hbit.rs:152]: These type parameters are unnecessary now.

File: cnd/src/swap_protocols/hbit.rs [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+            return Err(error);
+        }
+    }
+
+    Ok(())
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct HtlcParams<L, A, I> {
```
Comment (thomaseizinger) [inline cnd/src/swap_protocols/hbit.rs:197]: These are unnecessary as-well.

**Comment by tcharding** [comment]
Thanks for the review @thomaseizinger.

> I appreciate your effort @tcharding but I am a bit hesitant if this is not too much duplication now.

We have to kill rfc003 _somehow_. I don't see a better way than duplication.  It is much easier than trying to fix stuff at the same time, for example, how I left all the deploy stuff in even though hbit should not have deploy - IMO this will be easier to remove once the code is duplicated and isolated.

> Wouldn't it be easier if we would first finish ERC20:LNBTC and let the abstractions settle and sink in for a bit and also reduce some of the duplication?

I don't believe so, but there is more than one way to skin a cat.  We are struggling to parallelise the LN work now with things so entangled, the hlbit/herc20 stuff can at least be done in parallel.

> I have the feeling things become almost unmanageable if we add another 1500 LoC that are basically copied from somewhere in the codebase :(

I have the feeling, as I have aired before, that the codebase is unmanageable now, basically a total mess and we should rewrite the whole thing from scratch. You have previously suggested you do not want to do this. I'm at the end of my mind with how tangled everything is, at least this duplication allows us to work on an isolated part of the code. I'm not the architect but I'm the one of the wombats who has to struggle against this clusterf**k everyday :)

File: cnd/src/bitcoin.rs [hunk 5]
```diff
@@ -13,6 +13,50 @@ use serde::{
 };
 use std::{fmt, str::FromStr};
 
+/// The available Bitcoin networks.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum Network {
```
[the remark this answers is not in the document]
Comment (tcharding) [reply cnd/src/bitcoin.rs:18]: Well, since I was doing something totally new I thought I'd experiment with this - the three networks has been bothering me since we added them and no number of attempts at adding the Ethereum side has changed that (I've made 2 separate concerted efforts at it).

File: cnd/src/http_api/route_factory.rs [hunk 6]
```diff
@@ -169,9 +169,23 @@ pub fn create(
         .and(warp::path::param::<LocalSwapId>())
         .and(warp::path("refund"))
         .and(warp::path::end())
-        .and(facade)
+        .and(facade.clone())
         .and_then(http_api::routes::action_refund);
 
+    let hbit_herc20 = warp::post()
+        .and(warp::path!("swaps" / "hbit" / "herc20"))
```
[the remark this answers is not in the document]
Comment (tcharding) [reply cnd/src/http_api/route_factory.rs:176]: word.

File: cnd/src/http_api/routes/post.rs [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    let reply = warp::reply::reply();
+
+    // FIXME: Make this clone less ugly.
+    let cloned = body.clone();
+    let swap = CreatedSwap::<hbit::CreatedSwap, herc20::CreatedSwap> {
+        swap_id,
+        alpha: cloned.alpha.into(),
+        beta: cloned.beta.into(),
+        peer: cloned.peer.peer_id,
+        address_hint: None, // FIXME: Add this.
```
[the remark this answers is not in the document]
Comment (tcharding) [reply cnd/src/http_api/routes/post.rs:39]: That sounds good to me, will work this in.

File: cnd/src/network/comit_ln.rs [hunk 8]
```diff
@@ -628,14 +636,48 @@ impl NetworkBehaviourEventProcess<oneshot_behaviour::OutEvent<finalize::Message>
     }
 }
 
+#[async_trait]
+impl InitCommunication for ComitLN {
+    async fn init_communication<T>(&self, swap_id: LocalSwapId, params: T) -> anyhow::Result<()> {
+        // TODO: Implement generic communication methods based on
```
[the remark this answers is not in the document]
Comment (tcharding) [reply cnd/src/network/comit_ln.rs:642]: I tried that and got compiler problems because of `Send`/`'sync`.  Caused apparently by the MDNS, this current attempt was the second failed attempt, any more suggestions appreciated.

File: cnd/src/swap_protocols/facade.rs [hunk 9]
```diff
@@ -32,37 +37,56 @@ pub struct HanEtherereumHalightBitcoinCreateSwapParams {
...
+#[derive(Clone, Digest, Debug, PartialEq)]
+#[digest(hash = "SwapDigest")]
+pub struct Herc20HbitSwapParams {
+    #[digest(ignore)]
+    pub role: Role,
+    #[digest(ignore)]
+    pub peer: DialInformation,
+    #[digest(ignore)]
+    pub ethereum_identity: identity::Ethereum,
+    #[digest(prefix = "2001")]
```
[the remark this answers is not in the document]
Comment (tcharding) [reply cnd/src/swap_protocols/facade.rs:78]: Done: https://github.com/comit-network/comit-rs/issues/2624

File: cnd/src/swap_protocols/hbit.rs (hunk 1 above)
Comment (tcharding) [reply cnd/src/swap_protocols/hbit.rs:82]: Good catch, thanks.  I cut and pasted with no thought for the tracing - that is a habit I need to remedy :)

File: cnd/src/bitcoin.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply cnd/src/bitcoin.rs:18]: Fair, maybe that kind of use of the type system is too much :)
Do you think we should turn around on this topic and go back to enums?

**Comment by thomaseizinger** [comment]
> We have to kill rfc003 _somehow_. I don't see a better way than duplication. It is much easier than trying to fix stuff at the same time, for example, how I left all the deploy stuff in even though hbit should not have deploy - IMO this will be easier to remove once the code is duplicated and isolated.

I agree with killing it with duplication but I think as you duplicate stuff, you should at the same time make the changes to the code that you want to make (like removing `deploy`) otherwise there is no point in duplicating it, you might as-well use the old stuff.

> I have the feeling, as I have aired before, that the codebase is unmanageable now, basically a total mess and we should rewrite the whole thing from scratch. You have previously suggested you do not want to do this. I'm at the end of my mind with how tangled everything is, at least this duplication allows us to work on an isolated part of the code. I'm not the architect but I'm the one of the wombats who has to struggle against this clusterf**k everyday :)

Do you think it would help to take this as an opportunity to introduce a `comit` crate that holds all the protocols and btsieve and we slowly move things from cnd over and keep things "nice" over there? :)
This way, there would be a clear separation of code that we definitely want to keep, in a shape that we are happy with and the original modules in `cnd` are deprecated.

File: cnd/src/bitcoin.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (tcharding) [reply cnd/src/bitcoin.rs:18]: Yes, I do not think these three 'network' structs are a good use of the language features.  Also when we remove rfc003 I believe there original usecase is going to disappear (if I understood correctly why we added them in the first place). This PR shows that an enum serves us just nicely.

File: cnd/src/created_swap.rs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    pub alpha: A,
+    /// The parameters used on the beta ledger.
+    pub beta: B,
+    /// Peer ID of the swap counterparty.
+    pub peer: PeerId,
+    /// Role of the node in this swap, Alice or Bob.
+    pub role: Role,
+}
+
+pub type HbitHerc20CreatedSwap = CreatedSwap<hbit::CreatedSwap, herc20::CreatedSwap>;
```
Comment (luckysori) [inline cnd/src/created_swap.rs:24]: 🙃 Not too keen on this. Trading less characters for more indirection and basically the same name is not worth it for me.

File: cnd/src/db/created_swap.rs [hunk 11]
```diff
@@ -142,14 +137,131 @@ impl Load<CreatedSwap<herc20::CreatedSwap, halight::CreatedSwap>> for Sqlite {
...
+        let hbit = InsertableHbit {
+            swap_id: 0, // FK, set during save.
+            amount: Text(created.alpha.amount.into()),
+            network: Text(created.alpha.network),
+            hash_function: Text(HashFunction::Sha256),
+            redeem_identity: redeem_identity.map(|pk| Text(pk.into())),
+            refund_identity: refund_identity.map(|pk| Text(pk.into())),
+            ledger: Text(Ledger::Alpha),
+        };
+        self.save_hbit(local_swap_id, &hbit).await?;
```
Comment (luckysori) [inline cnd/src/db/created_swap.rs:178]: 🤔 Could be extracted into a function to reduce duplication with function below. Or at least scoped. That way the comment on line 159 can go away too.

Comment (luckysori) [reply cnd/src/db/created_swap.rs:178]: Same for saving herc20 details.

File: cnd/src/http_api.rs [hunk 12]
```diff
@@ -375,9 +352,9 @@ pub enum HttpAssetParams {
     Erc20(Erc20AssetParams),
 }
 
-#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq)]
 pub struct BitcoinAssetParams {
-    quantity: Http<asset::Bitcoin>,
```
Comment (luckysori) [inline cnd/src/http_api.rs:380]: ❓ Why the move away from the custom deserialisation?

File: cnd/src/http_api/route_factory.rs [hunk 13]
```diff
@@ -133,6 +133,14 @@ pub fn create(
         .and(facade.clone())
         .and_then(http_api::routes::index::post_halight_bitcoin_herc20);
 
+    // This will replace get_halight_swap once implemented.
```
Comment (luckysori) [inline cnd/src/http_api/route_factory.rs:136]: 🔧 Seems obvious, could just remove these comments?

File: cnd/src/http_api/route_factory.rs [hunk 14]
```diff
@@ -198,6 +220,9 @@ pub fn create(
         .or(lightning_action_deploy)
         .or(lightning_action_redeem)
         .or(lightning_action_refund)
+        .or(hbit_herc20)
+        .or(herc20_hbit)
+        //        .or(get_swaps)
```
Comment (luckysori) [inline cnd/src/http_api/route_factory.rs:225]: 🔧 Would remove.

File: cnd/src/http_api/routes/post.rs [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    db::Save,
+    http_api::{problem, routes::into_rejection, Http},
+    identity,
+    network::{DialInformation, InitCommunication},
+    swap_protocols::{hbit, herc20, Facade, LocalSwapId, Role},
+};
+use serde::Deserialize;
+use warp::{http::StatusCode, Rejection, Reply};
+
+/// POST endpoints for the hbit/herc20 protocol pair.
```
Comment (luckysori) [inline cnd/src/http_api/routes/post.rs:13]: ❓ Instead of having to clarify with this documentation string, should we just call the file something like `post_hbit_herc20.rs`. I guess eventually this all moves back to `index.rs` anyway.

Comment (luckysori) [reply cnd/src/http_api/routes/post.rs:13]: ❓ I don't immediately see why all this is not in `index.rs` already anyway. How does this duplication help us?

File: cnd/src/swap_protocols/facade.rs [hunk 16]
```diff
@@ -34,6 +38,96 @@ pub struct Herc20HalightBitcoinCreateSwapParams {
     pub lightning_amount: asset::Bitcoin,
 }
 
+/// This represent the information available on a swap
```
Comment (luckysori) [inline cnd/src/swap_protocols/facade.rs:41]: 🔧 
```suggestion
/// This represents the information available on a swap
```

Comment (luckysori) [reply cnd/src/swap_protocols/facade.rs:41]: `Herc20HbitSwapParams` below is missing a similar docstring.

File: cnd/src/swap_protocols/hbit.rs [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+use genawaiter::{
+    sync::{Co, Gen},
+    GeneratorState,
+};
+use std::sync::Arc;
+use tracing_futures::Instrument;
+
+/// Htlc Bitcoin atomic swap protocol.
+
+/// Data required to create a swap that involves Ether.
```
Comment (luckysori) [inline cnd/src/swap_protocols/hbit.rs:41]:
```suggestion
/// Data required to create a swap that involves Bitcoin.
```

File: cnd/src/swap_protocols/hbit/ledger_states.rs [hunk 18]
```diff
@@ -0,0 +1,84 @@
+use crate::swap_protocols::{
+    hbit::{LedgerState, SwapEvent},
+    LocalSwapId,
+};
+use async_trait::async_trait;
+use std::collections::HashMap;
+use tokio::sync::Mutex;
+
+#[async_trait]
+pub trait Insert: Send + Sync + 'static {
```
Comment (luckysori) [inline cnd/src/swap_protocols/hbit/ledger_states.rs:10]: ❓ I could just be very far behind in terms of this codebase, but why are we not using the traits provided by `swap_protocols/state.rs`?

File: cnd/src/db/tables.rs [hunk 19]
```diff
@@ -189,6 +189,46 @@ impl InsertableHalight {
     }
 }
 
+#[derive(Associations, Clone, Copy, Debug, Identifiable, Queryable, PartialEq)]
+#[belongs_to(Swap)]
+#[table_name = "hbits"]
+pub struct Hbit {
+    id: i32,
+    swap_id: i32,
+    pub amount: Text<Satoshis>, // FIXME: Do we want to store this as Satoshis?
```
Comment (bonomat) [inline cnd/src/db/tables.rs:198]: Would the alternative be to store it in Bitcoin? 
I think storing Satoshi makes sense to not have to convert Satoshi to Bitcoin and not use floating point numbers.

File: cnd/src/db/tables.rs [hunk 20]
```diff
@@ -189,6 +189,46 @@ impl InsertableHalight {
...
+    pub network: Text<bitcoin::Network>,
+    pub hash_function: Text<HashFunction>,
+    pub redeem_identity: Option<Text<bitcoin::PublicKey>>,
+    pub refund_identity: Option<Text<bitcoin::PublicKey>>,
+    pub ledger: Text<Ledger>,
+}
+
+#[derive(Insertable, Clone, Copy, Debug)]
+#[table_name = "hbits"]
+pub struct InsertableHbit {
```
Comment (bonomat) [inline cnd/src/db/tables.rs:208]: Is this best-practice? 
I'd assume you can reuse the same type for inserting and loading.

File: cnd/src/http_api/routes.rs [hunk 21]
```diff
@@ -202,7 +204,6 @@ pub async fn handle_get_halight_swap(
     }
 }
 
-// TODO: Refactor with make_finalized_swap_entity
```
Comment (bonomat) [inline cnd/src/http_api/routes.rs:205]: Instead of describing what someone needs to do, can you shortly explain why this should be refactored in this manner please? 
This might also indeed be better stored in a ticket as otherwise this will get lost here.

File: cnd/src/http_api/routes/get.rs [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+    swap_protocols::{Facade, LocalSwapId},
+};
+use http_api_problem::HttpApiProblem;
+use warp::{Rejection, Reply};
+
+pub fn into_rejection(problem: HttpApiProblem) -> Rejection {
+    warp::reject::custom(problem)
+}
+
+#[allow(clippy::needless_pass_by_value)]
```
Comment (bonomat) [inline cnd/src/http_api/routes/get.rs:12]: Is my understanding correct that this will go away once we actually do the loading?

**Comment by tcharding** [comment]
> I think it's okay to not tackle the unnecessary deploy of hbit (is that the new name?) with this PR, since I believe we still have it in rfc003. It's not like it's a step backwards. I would just remove those TODOs, we already know that we want that step to go away.

Cool, will do.  Thanks.  Oh and yes this is the new name bestowed upon us by the Gods i.e., me :)

File: cnd/src/created_swap.rs (hunk 10 above)
Comment (tcharding) [reply cnd/src/created_swap.rs:24]: ok, I was indecisive about theses aliases, will remove.

File: cnd/src/http_api/routes/post.rs [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+            warp::reply::with_status(
+                warp::reply::with_header(reply, "Location", format!("/swaps/{}", swap_id)),
+                StatusCode::CREATED,
+            )
+        })
+        .map_err(problem::from_anyhow)
+        .map_err(into_rejection)
+}
+
+#[allow(clippy::needless_pass_by_value)]
```
Comment (bonomat) [inline cnd/src/http_api/routes/post.rs:52]: Is this still valid?

File: cnd/src/db/created_swap.rs (hunk 11 above)
Comment (tcharding) [reply cnd/src/db/created_swap.rs:178]: Nice, will do.

File: cnd/src/http_api.rs (hunk 12 above)
Comment (tcharding) [reply cnd/src/http_api.rs:380]: This was likely a mistake by me, have discussed with @thomaseizinger.  Am planning on removing this from the PR.

File: cnd/src/http_api/route_factory.rs (hunk 13 above)
Comment (tcharding) [reply cnd/src/http_api/route_factory.rs:136]: Oh, you thought that was obvious?  I thought `get_halight_swap` was so mis-appropriately named that it was not obvious at all :)

File: cnd/src/http_api/route_factory.rs (hunk 14 above)
Comment (tcharding) [reply cnd/src/http_api/route_factory.rs:225]: Can do.

File: cnd/src/http_api/routes/post.rs (hunk 15 above)
Comment (tcharding) [reply cnd/src/http_api/routes/post.rs:13]: > Instead of having to clarify with this documentation string, should we just call the file something like post_hbit_herc20.rs.

I imagine _all_ the POST funtions to be in here.  This comment, that risks going stale, was to make it apparent that so far we only have this trading pair done - it feels like the `http_api/` is going to be hard to navigate because the naming of rfc003 files (`index.rs` etc.) is not being used now so its hard to tell which files are which.

> I don't immediately see why all this is not in index.rs already anyway. How does this duplication help us?

`index.rs` is going away, if I get my way, for 2 reasons. 1. because its not a useful name 2. Because it mostly contains halight code that needs rewriting. I did not want to break the LN code though so did not delete it yet.

File: cnd/src/swap_protocols/facade.rs (hunk 16 above)
Comment (tcharding) [reply cnd/src/swap_protocols/facade.rs:41]: Lol, I egotistically did not fix this because I don't like these structs :)  Will fix.

File: cnd/src/swap_protocols/hbit.rs (hunk 17 above)
Comment (tcharding) [reply cnd/src/swap_protocols/hbit.rs:41]: Nice, thanks.

File: cnd/src/swap_protocols/hbit/ledger_states.rs (hunk 18 above)
Comment (tcharding) [reply cnd/src/swap_protocols/hbit/ledger_states.rs:10]: Good and fair point, I will investigate.  Discussed already with @da-kami and @thomaseizinger and will either
- Use traits in `swap_protocols/state.rs`
- Remove the traits and just have methods.

File: cnd/src/swap_protocols/facade.rs [hunk 24]
```diff
@@ -87,14 +166,12 @@ pub struct Facade {
     pub swarm: Swarm,
     // We currently only support Han-HALight, therefor 'alpha' is Ethereum and 'beta' is Lightning.
     pub alpha_ledger_states: Arc<LedgerStates>,
-    pub beta_ledger_states: Arc<halight::States>,
+    pub halight_states: Arc<halight::States>,
```
Comment (bonomat) [inline cnd/src/swap_protocols/facade.rs:169]: Why was this renamed?

File: cnd/src/swap_protocols/hbit.rs [hunk 25]
```diff
@@ -0,0 +1,204 @@
...
+pub struct HtlcParams {
+    pub network: bitcoin::Network,
+    pub asset: asset::Bitcoin,
+    pub redeem_identity: identity::Bitcoin,
+    pub refund_identity: identity::Bitcoin,
+    pub expiry: Timestamp,
+    pub secret_hash: SecretHash,
+}
+
+// TODO: Remove the Herc20 stuff i.e., no separate Deployed/Funded states.
```
Comment (bonomat) [inline cnd/src/swap_protocols/hbit.rs:174]: This should be tracked somewhere else. 
I propose to add a comment above at around line 135 explaining why we have this right now and creating a ticket to remove the unneeded swap events.

File: cnd/src/swap_protocols/hbit/events.rs [hunk 26]
```diff
@@ -0,0 +1,77 @@
+use crate::{
+    asset, htlc_location,
+    swap_protocols::{hbit::HtlcParams, Secret},
+    transaction,
+};
+use chrono::NaiveDateTime;
+use serde::{Deserialize, Serialize};
+
+// TODO: Remove the Herc20 stuff i.e., no separate Deployed/Funded states.
```
Comment (bonomat) [inline cnd/src/swap_protocols/hbit/events.rs:9]: This TODO can go away if you resolve my comment in `cnd/src/swap_protocols/hbit.rs`

**Comment by D4nte** [comment]
Lots of detailed reviews happening already. Not sure my input will help.
Please note I am confused by `hbit` name as it could both mean `halightbitcoin` and `hanbitcoin`.

File: cnd/src/bitcoin.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bonomat) [reply cnd/src/bitcoin.rs:18]: Why is this introduced now and was not raised earlier? 

I'm afraid that introducing this here now means for sure that we have to refactor one or the other in the future. 
I'm not against this but we should stick to one solution and not jump back and forth.

File: cnd/src/swap_protocols/hbit/ledger_states.rs [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+                tracing::warn!("Value not found for key {}", key);
+                return;
+            }
+        };
+
+        match event {
+            SwapEvent::Deployed(deployed) => ledger_state.transition_to_deployed(deployed),
+            SwapEvent::Funded(funded) => ledger_state.transition_to_funded(funded),
+            SwapEvent::Redeemed(redeemed) => {
+                // what if redeemed.secret.hash() != secret_hash in request ??
```
Comment (bonomat) [inline cnd/src/swap_protocols/hbit/ledger_states.rs:61]: Can you clarify how this can happen?

**Comment by tcharding** [comment]
> Lots of detailed reviews happening already. Not sure my input will help.
> Please note I am confused by `hbit` name as it could both mean `halightbitcoin` and `hanbitcoin`.

'h' stands for HTLC, 'bit' stands for Bitcoin. `halight` will be relegated to our memories soon as we can come up with a name with at least a small amount of consensus, most likely `halbit`.

File: cnd/src/bitcoin.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply cnd/src/bitcoin.rs:18]: > I'm afraid that introducing this here now means for sure that we have to refactor one or the other in the future.

We already have to do that because `Ethereum` is still using an enum :D
I am fine with moving back on this one, was a bad idea.

File: cnd/src/db/tables.rs (hunk 20 above)
Comment (thomaseizinger) [reply cnd/src/db/tables.rs:208]: They differ in the way that `Insertable` doesn't have a primary key. Not sure if you can reuse them then?

File: cnd/src/http_api/routes/get.rs (hunk 22 above)
Comment (thomaseizinger) [reply cnd/src/http_api/routes/get.rs:12]: No I don' think so. The problem is that `warp` passes us this by value so we actually cannot change the fn signature but clippy doesn't know that :(

File: cnd/src/db/created_swap.rs (hunk 11 above)
Comment (tcharding) [reply cnd/src/db/created_swap.rs:178]: Had a play with this; it actually got me to somewhere I wanted to be without previously know how :)

We can add another type that groups with all the insertable/queryable types i.e.,
- `Foo`, `InsertableFoo`, and `FooData`

Open to a better name than `FooData`, the obvious is to use `Foo` and use `QueryableFoo` for the queryable type but this is out because of how the `belongs_to` macro works, well at least for the `Swap` type, and I'd like to be uniform.

Anyways, leaving this for a follow up PR because it touches all the db code.

File: cnd/src/db/tables.rs (hunk 20 above)
Comment (bonomat) [reply cnd/src/db/tables.rs:208]: It won't work to make this field optional?

File: cnd/src/db/created_swap.rs (hunk 11 above)
Comment (tcharding) [reply cnd/src/db/created_swap.rs:178]: What you have touched on here @lucas is actually a really big deal, fixing this makes the whole db code better. Easier to read, write, extend, and maintain. I was looking for this, thanks.

https://github.com/comit-network/comit-rs/issues/2635

File: cnd/src/swap_protocols/hbit/ledger_states.rs (hunk 18 above)
Comment (tcharding) [reply cnd/src/swap_protocols/hbit/ledger_states.rs:10]: Removed the traits. Thanks.

File: cnd/src/http_api.rs (hunk 12 above)
Comment (tcharding) [reply cnd/src/http_api.rs:380]: So the upside of this is: left the custom serder in `bitcoin` module (because it is used by `strum_discriminants`/`HtlcState`), but did not use it in the http_api, instead reverted to using custom serde and wrapping in `Http` type. The `bitcoin` module serde stuff should be removed when we remove the `strum_discriminants` usage to create `HtlpState` (this relies on all types being serde).

Tracking issue: https://github.com/comit-network/comit-rs/issues/2636

File: cnd/src/asset/bitcoin.rs [hunk 28]
```diff
@@ -31,6 +32,29 @@ impl fmt::Display for Bitcoin {
...
+        D: Deserializer<'de>,
+    {
+        let value = String::deserialize(deserializer)?;
+        let value =
+            u64::from_str(value.as_str()).map_err(<D as Deserializer<'de>>::Error::custom)?;
+        let amount = Bitcoin::from_sat(value);
+
+        Ok(amount)
+    }
+}
```
[the remark this answers is not in the document]
Comment (tcharding) [reply cnd/src/asset/bitcoin.rs:56]: Seen to in above comments, along with tracking issue.

File: cnd/src/swap_protocols/hbit/events.rs (hunk 26 above)
Comment (tcharding) [reply cnd/src/swap_protocols/hbit/events.rs]: Removed, thanks.

File: cnd/src/swap_protocols/hbit/ledger_states.rs (hunk 27 above)
Comment (tcharding) [reply cnd/src/swap_protocols/hbit/ledger_states.rs:38]: Ahhh, funny that should come up now - this is one of those places in the code that should have the label `FIXME`. I did not write this comment and have no clue what it means, hence I'd suggest we remove the CI rule and encourage devs to use
`FIXME (tcharding): blah blah`

Comment (tcharding) [reply cnd/src/swap_protocols/hbit/ledger_states.rs:38]: > Why is this introduced now and was not raised earlier?

It is introduced now because this PR does a bunch of code duplication that uses the structs that this enum replaces.

> I'm afraid that introducing this here now means for sure that we have to refactor one or the other in the future.

We definitely have a *lot* of refactoring to do to remove all the rfc003 kruft.

> I'm not against this but we should stick to one solution and not jump back and forth.

The stated aim of this PR was to write clean code _without_ breaking anyone else's code, hence we now have multiple abstractions for the same thing. (There is also an enum in the db layer that represents this also :)

File: cnd/src/db/tables.rs (hunk 20 above)
Comment (da-kami) [reply cnd/src/db/tables.rs:208]: Please don'e try. We wasted enough time on trying to do such kind of optimisations. 

@bonomat we are not the only ones [discussing](https://www.reddit.com/r/rust/comments/9r0n1e/diesel_queryableinsertable/) this, but as far as I could see the typical way is still to use 2 different structs for insert and query.

File: cnd/src/http_api/routes.rs (hunk 21 above)
Comment (D4nte) [reply cnd/src/http_api/routes.rs:205]: I Wrote this todo and actioned it. I wonder how it go in dev.

File: cnd/src/db/created_swap.rs [hunk 29]
```diff
@@ -142,14 +137,131 @@ impl Load<CreatedSwap<herc20::CreatedSwap, halight::CreatedSwap>> for Sqlite {
...
+            Role::Alice => Some(created.alpha.identity),
+            Role::Bob => None,
+        };
+
+        let hbit = InsertableHbit {
+            swap_id: 0, // FK, set during save.
+            amount: Text(created.alpha.amount.into()),
+            network: Text(created.alpha.network),
+            hash_function: Text(HashFunction::Sha256),
+            redeem_identity: redeem_identity.map(|pk| Text(pk.into())),
```
[the remark this answers is not in the document]
Comment (tcharding) [reply cnd/src/db/created_swap.rs:174]: @lucas gave me an even better idea to clean all this constructor code up. Have opened a ticket for it so as to not spin for too long on perfecting the code. Thanks for the review.

File: cnd/src/swap_protocols/hbit/ledger_states.rs (hunk 27 above)
Comment (thomaseizinger) [reply cnd/src/swap_protocols/hbit/ledger_states.rs:38]: @bonomat You commented on quite a few things that were not originally written in this PR but just copied over. I am not sure how helpful that is here :grimacing: 

We can debate how carefully one should be when copying code. I would argue quite careful but in this case, the main aim was to have a clean slate to work off from so I don't mind much :)

Comment (bonomat) [reply cnd/src/swap_protocols/hbit/ledger_states.rs:38]: How should I know this? It is green and marked as new, hence, it asked for a review ¯\_(ツ)_/¯ 

None of my comments are blocking, so don't worry too much about this.

Comment (thomaseizinger) [reply cnd/src/swap_protocols/hbit/ledger_states.rs:38]: > How should I know this? It is green and marked as new, hence, it asked for a review ¯_(ツ)_/¯

Fair point :)

File: cnd/src/swap_protocols/facade.rs (hunk 24 above)
Comment (tcharding) [reply cnd/src/swap_protocols/facade.rs:169]: Because the name is wrong, its an artifact of rfc003.
